webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+
+## Motivation
+
+1. **No need for trusted TLS certificates.** Enable browsers to connect to
+   public server nodes without those server nodes providing a TLS certificate
+   within the browsers trustchain. Note that we can not do this today with our
+   Websocket transport as the browser requires the remote to have a trusted TLS
+   certificate. Nor can we establish a plain TCP or QUIC connection from within
+   a browser.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   server node to connect even though one or both are behind a NAT / firewall.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Addressing
+
+WebRTC multiaddresses are composed of an IP and UDP address component, followed
+by `/webrtc` and a multihash of the certificate that the node uses.
+
+Examples:
+- `/ip4/1.2.3.4/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+- `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+
+The TLS certificate fingerprint in `/certhash` is a
+[multibase](https://github.com/multiformats/multibase) encoded
+[multihash](https://github.com/multiformats/multihash).
+
+For compatibility implementations MUST support hash algorithm
+[`sha-256`](https://github.com/multiformats/multihash) and base encoding
+[`base64url`](https://github.com/multiformats/multibase). Implementations MAY
+support other hash algorithms and base encodings, but they may not be able to
+connect to all other nodes.
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+As a preparation browser _A_ [generates a
+certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate)
+and [gets the certificate's
+fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ instantiates a `RTCPeerConnection`, passing its local certificate as a
+   parameter. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ creates a local offer via
+   [`RTCPeerConnection.createOffer()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createOffer).
+   _A_ generates a random string and sets that string as the username (_ufrag_
```

Comment:
https://www.rfc-editor.org/rfc/rfc5245#section-15.4

min: 22 characters
max: 256 characters
ALPHA / DIGIT / "+" / "/"
at least 128 bits of randomness

Reply:
Given the requirements above through the WebRTC specification, what would be the value of specifying stricter rules in the libp2p WebRTC specification?

We could require a certain length for the client ufrag to increase the size of the client STUN message and thus decrease the impact of an amplification attack as previously suggested by @marten-seemann.

For now, we settled on not including this mechanism, and instead assume the QUIC methodology to be good enough.

https://github.com/libp2p/specs/pull/412/files#diff-10725e780e3c8583b65c128ab23ebc6b3f5502f68b922e33bd86d941c7d601c3R130-R142

Let me know in case you would still like to further restrict these values.

Reply:
> For now, we settled on not including this mechanism, and instead assume the QUIC methodology to be good enough.

Can you clarify which QUIC methodology you’re referring to? We’re using neither the minimum initial packet size nor do we define an amplification factor in this spec, do we?
If I understand correctly, our DoS defense here is that - ignoring packet sizes - the server will only send a single packet in response to each packet received from the client.

Reply:
> Can you clarify which QUIC methodology you’re referring to?

I was referring to:

> an endpoint MUST limit the amount of data it sends to the unvalidated
> address to three times the amount of data received from that address.

https://datatracker.ietf.org/doc/html/rfc9000#section-8

That said, it is still an open item here to validate that the above is actually the case for WebRTC.

Assuming this is the case for WebRTC, do you still see the need to further restrict the ufrag values @marten-seemann?

Reply:
I am interpreting the silence as agreement to not needing to further restrict the length. Thus resolving here.